### PR TITLE
Tests should produce less false positives

### PR DESCRIPTION
**Pull request by mtrna**
Tests should produce less false positives *without intervention in the environment* by:
* ~~increasing time interval in a `MetricManagerTest`~~
* fixing the source for computing of the time interval in a `MetricManagerTest`
* closing file descriptors in a `DirectoryFlowLoaderTest`

Also, some typos were fixed.

This MR should not have any functional impact. Goal is to improve development experience.

Consider merging https://github.com/azkaban/azkaban/pull/1647, please. Thanks! :)

### CODE REVIEW

File: azkaban-common/src/test/java/azkaban/metric/MetricManagerTest.java [hunk 1]
```diff
@@ -81,12 +81,12 @@ public void managerEmitterMaintenanceTest() {
...
   public void managerEmitterHandlingTest() throws Exception {
     this.emitter.purgeAllData();
-    final Date from = DateTime.now().minusMinutes(1).toDate();
+    final Date from = DateTime.now().minusMinutes(10).toDate();
     this.metric.notifyManager();
 
     this.emitterWrapper.countDownLatch.await(10L, TimeUnit.SECONDS);
 
-    final Date to = DateTime.now().plusMinutes(1).toDate();
+    final Date to = DateTime.now().plusMinutes(10).toDate();
```
Comment (juhoautio) [inline azkaban-common/src/test/java/azkaban/metric/MetricManagerTest.java]: Do you know somehow that this was the culprit? Even 1 minute is a lot, so I would expect the bug to lie somewhere else.

**Comment by juhoautio** [comment]
See also https://github.com/azkaban/azkaban/issues/1725

File: azkaban-common/src/test/java/azkaban/metric/MetricManagerTest.java (hunk 1 above)
Comment (mtrna) [reply azkaban-common/src/test/java/azkaban/metric/MetricManagerTest.java]: Sure, that is an interesting question. :) I posed that to myself and the result you can see in the updated PR. :)

File: azkaban-common/src/test/java/azkaban/executor/ExecutionFlowDaoTest.java [hunk 2]
```diff
@@ -170,6 +170,9 @@ public void testFetchEmptyRecentlyFinishedFlows() throws Exception {
     final List<ExecutableFlow> flows = this.executionFlowDao
         .fetchRecentlyFinishedFlows(RECENTLY_FINISHED_LIFETIME);
     assertThat(flows.size()).isEqualTo(0);
+
+    //Restore the clock
+    DateTimeUtils.setCurrentMillisOffset(0);
```
Comment (juhoautio) [inline azkaban-common/src/test/java/azkaban/executor/ExecutionFlowDaoTest.java]: Oh, so this test leaks DateTimeUtils offset? It would be safer to place the reset in an `@After` method to not cause side effects even if the test case fails.

File: azkaban-common/src/test/java/azkaban/metric/MetricManagerTest.java [hunk 3]
```diff
@@ -80,13 +80,18 @@ public void managerEmitterMaintenanceTest() {
...
   @Test
   public void managerEmitterHandlingTest() throws Exception {
+
+    // metrics use System.currentTimeMillis, so that method should be the millis provider
+    final DateTime aboutNow = new DateTime(System.currentTimeMillis());
+
     this.emitter.purgeAllData();
-    final Date from = DateTime.now().minusMinutes(1).toDate();
+
+    final Date from = aboutNow.minusMinutes(1).toDate();
```
Comment (juhoautio) [inline azkaban-common/src/test/java/azkaban/metric/MetricManagerTest.java]: It should be ok to use the "simple" `DateTime.now()` as long DateTimeUtils is not in some bad state?

File: azkaban-common/src/test/java/azkaban/project/DirectoryFlowLoaderTest.java [hunk 4]
```diff
@@ -95,11 +96,14 @@ private static File decompressTarBZ2(InputStream is) throws IOException {
           parent.mkdirs();
         }
 
-        IOUtils.copy(tais, new FileOutputStream(outputFile));
+        FileOutputStream os = new FileOutputStream(outputFile);
+        IOUtils.copy(tais, os);
+        os.close();
```
Comment (juhoautio) [inline azkaban-common/src/test/java/azkaban/project/DirectoryFlowLoaderTest.java]: Should be in a separate commit, maybe even PR of its own. If this change even matters..

File: azkaban-common/src/test/java/azkaban/metric/MetricManagerTest.java (hunk 3 above)
Comment (juhoautio) [reply azkaban-common/src/test/java/azkaban/metric/MetricManagerTest.java]: I tried this – this makes the test fail consistently:

```java
  @Test
  public void managerEmitterHandlingTest() throws Exception {
    // ADDING THESE TWO LINES:
    final Duration FLOW_FINISHED_TIME = Duration.ofMinutes(2);
    DateTimeUtils.setCurrentMillisOffset(-FLOW_FINISHED_TIME.toMillis());

    this.emitter.purgeAllData();
    final Date from = DateTime.now().minusMinutes(1).toDate();
    this.metric.notifyManager();

    this.emitterWrapper.countDownLatch.await(10L, TimeUnit.SECONDS);
    assertEquals(0, this.emitterWrapper.countDownLatch.getCount());

    final Date to = DateTime.now().plusMinutes(1).toDate();
    final List<InMemoryHistoryNode> nodes = this.emitter.getMetrics("FakeMetric", from, to, false);

    assertEquals("Failed to report metric", 1, nodes.size());
    assertEquals("Failed to report metric", nodes.get(0).getValue(), 4);
  }
```

If I add this after the two added lines it works again:

```java
    DateTimeUtils.setCurrentMillisSystem();
```

My suggestion for the eventual fix would be to only modify `ExecutionFlowDaoTest.java` by adding this into its existing `@After` method:

```java
    DateTimeUtils.setCurrentMillisSystem();
```

I found one more leaking test which is `BasicTimeCheckerTest` – it should also reset DateTimeUtils in an `@After` method.

**Comment by juhoautio** [comment]
Great find @mtrna !

**Comment by juhoautio** [comment]
Btw, how did you find this? Were you able to somehow run the test locally so that it failed consistently (so that you could debug what the actual timestamps are)?

File: azkaban-common/src/test/java/azkaban/project/DirectoryFlowLoaderTest.java (hunk 4 above)
Comment (mtrna) [reply azkaban-common/src/test/java/azkaban/project/DirectoryFlowLoaderTest.java]: Hi @juhoautio ! It matters, since it allows to release the file descriptors. `IOException` thrown upon the exhaustion of file descriptors is the reason why this test frequently fails.

File: azkaban-common/src/test/java/azkaban/executor/ExecutionFlowDaoTest.java (hunk 2 above)
Comment (mtrna) [reply azkaban-common/src/test/java/azkaban/executor/ExecutionFlowDaoTest.java]: fair point, I will update the PR

File: azkaban-common/src/test/java/azkaban/metric/MetricManagerTest.java (hunk 3 above)
Comment (mtrna) [reply azkaban-common/src/test/java/azkaban/metric/MetricManagerTest.java]: `DateTime.now()` takes the milliseconds from the effective joda millis provider [1]. Since metrics use the system millis provider directly [citation needed :D], I suggest the `MetricManagerTest` use the system millis provider as well.

[1] https://github.com/JodaOrg/joda-time/blob/master/src/main/java/org/joda/time/DateTimeUtils.java#L112

File: azkaban-common/src/test/java/azkaban/project/DirectoryFlowLoaderTest.java (hunk 4 above)
Comment (mtrna) [reply azkaban-common/src/test/java/azkaban/project/DirectoryFlowLoaderTest.java]: I agree, will update the PR later today

File: azkaban-common/src/test/java/azkaban/metric/MetricManagerTest.java (hunk 3 above)
Comment (mtrna) [reply azkaban-common/src/test/java/azkaban/metric/MetricManagerTest.java]: joda `DateTime` and `minus/plusMinutes` are used, though! :)

**Comment by mtrna** [comment]
updated, rebased; PR is ready for another round of review

**Comment by HappyRay** [comment]
I haven't had time to look at this PR in details yet. 
@mtrna I appreciate your contribution to make our product better, based on your description!

I am glad to see the community-driven reviews. 
Thanks @juhoautio !

**Comment by juhoautio** [comment]
How about my suggestion on fixing this also in BasicTimeCheckerTest?

File: azkaban-common/src/test/java/azkaban/metric/MetricManagerTest.java (hunk 3 above)
Comment (juhoautio) [reply azkaban-common/src/test/java/azkaban/metric/MetricManagerTest.java]: Yeah it just adds one line of code more. To me there’s no difference if using joda time directly because underneath it should also be calling System#currentTimeMillis. Also it’s safer to get the current time again _after_ the emitter has been called. This shouldn’t really happen, but if the metric call would take more that one minute this could fail because the to date is not after that.

**Comment by mtrna** [comment]
added a fix of the `BasicTimeCheckerTest`

File: azkaban-common/src/test/java/azkaban/executor/ExecutionFlowDaoTest.java [hunk 5]
```diff
@@ -357,4 +357,12 @@ private void assertTwoFlowSame(final ExecutableFlow flow1, final ExecutableFlow
     assertThat(new HashSet<>(flow1.getEndNodes())).isEqualTo(new HashSet<>(flow2.getEndNodes()));
   }
 
+  /**
+   * restores the clock; see {@link #testFetchEmptyRecentlyFinishedFlows()}
+   */
+  @After
+  public void clockReset() {
+    DateTimeUtils.setCurrentMillisOffset(0);
```
Comment (juhoautio) [inline azkaban-common/src/test/java/azkaban/executor/ExecutionFlowDaoTest.java]: I checked DateTimeUtils#setCurrentMillisSystem – calling setCurrentMillisOffset(0) is effectively the same. So why not use the simpler method that doesn't require an arg?

File: azkaban-common/src/test/java/azkaban/metric/MetricManagerTest.java [hunk 6]
```diff
@@ -80,13 +80,18 @@ public void managerEmitterMaintenanceTest() {
...
     this.metric.notifyManager();
 
     this.emitterWrapper.countDownLatch.await(10L, TimeUnit.SECONDS);
 
-    final Date to = DateTime.now().plusMinutes(1).toDate();
+    final Date to = aboutNow.plusMinutes(1).toDate();
```
Comment (juhoautio) [inline azkaban-common/src/test/java/azkaban/metric/MetricManagerTest.java]: Should get the current time again here instead of using `aboutNow` from before the `notifyManager()` call.

**Comment by juhoautio** [comment]
Left 2 more comments, otherwise 👍 from my perspective.

File: azkaban-common/src/test/java/azkaban/executor/ExecutionFlowDaoTest.java (hunk 5 above)
Comment (mtrna) [reply azkaban-common/src/test/java/azkaban/executor/ExecutionFlowDaoTest.java]: reuse of existing code; but sure we can replace it as well :)

File: .gitignore [hunk 7]
```diff
@@ -26,3 +26,5 @@ temp/
 .shelf
 # Intellij default build output directory
 out/
+
+currentpid
```
Comment (HappyRay) [inline .gitignore]: A new line is needed at the end.

Did you follow the contributor guide and set up Save actions plugin?

File: azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java [hunk 8]
```diff
@@ -1734,7 +1734,7 @@ private void cleanExecutionLogs() {
 
     private static final long QUEUE_PROCESSOR_WAIT_IN_MS = 1000;
     private final int maxDispatchingErrors;
-    private final long activeExecutorRefreshWindowInMilisec;
+    private final long activeExecutorRefreshWindowInMillisec;
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java]: Thanks for fixing typos. 
In the future, I would suggest separate PRs for independent changes.

**Comment by juhoautio** [comment]
> @juhoautio do you think this change is going to fix the recent unit test in-stability reported by multiple users?

Yes.

> I trust @juhoautio's judgment. If he is ok with it, chances are the change is fine. :)

👍 from me for this PR in its current state.

**Comment by mtrna** [comment]
resolved conflicts with master

File: azkaban-common/src/test/java/azkaban/executor/ExecutionFlowDaoTest.java (hunk 5 above)
Comment (kunkun-tang) [inline azkaban-common/src/test/java/azkaban/executor/ExecutionFlowDaoTest.java]: In the class `DefaultMailCreatorTest`,

https://github.com/azkaban/azkaban/blob/c75ea9beaea61ef3d92a2948b0d16834e078fb1f/azkaban-common/src/test/java/azkaban/executor/mail/DefaultMailCreatorTest.java#L53

Should we also change DataTimeUtils back when that test finishes?

Comment (juhoautio) [reply azkaban-common/src/test/java/azkaban/executor/ExecutionFlowDaoTest.java]: No need to change it, it's already done properly:
https://github.com/azkaban/azkaban/blob/c75ea9beaea61ef3d92a2948b0d16834e078fb1f/azkaban-common/src/test/java/azkaban/executor/mail/DefaultMailCreatorTest.java#L84

Comment (kunkun-tang) [reply azkaban-common/src/test/java/azkaban/executor/ExecutionFlowDaoTest.java]: Thanks for checking!
